rosbag2_compression/src/rosbag2_compression/zstd_decompressor.cpp

```diff
@@ -0,0 +1,201 @@
+{
+  FILE * fp{nullptr};
+#ifdef _WIN32
+  fopen_s(&fp, uri.c_str(), read_mode.c_str());
+#else
+  fp = std::fopen(uri.c_str(), read_mode.c_str());
+#endif
+  return fp;
+}
+
+std::vector<uint8_t> get_input_buffer(
+  const std::string & uri,
+  size_t & compressed_buffer_length)
+{
+  // Get the file size
+  compressed_buffer_length = rosbag2_storage::FilesystemHelper::get_file_size(uri);
+
+  // Read in buffer, handling accordingly
+  const auto file_pointer = open_file(uri.c_str(), "rb");
+  if (file_pointer == nullptr) {
+    std::stringstream errmsg;
+    errmsg << "Failed to open file: \"" << uri << "\" for binary reading!";
+
+    throw std::runtime_error(errmsg.str());
+  }
+
+  // Allocate and read in
+  std::vector<uint8_t> compressed_buffer;
+  compressed_buffer.reserve(compressed_buffer_length);
+
+  const auto nRead = fread(
+    compressed_buffer.data(), sizeof(uint8_t), compressed_buffer_length, file_pointer);
+
+  if (nRead != compressed_buffer_length) {
+    ROSBAG2_COMPRESSION_LOG_ERROR_STREAM("Bytes read !(" <<
+      nRead << ") != compressed_buffer_length (" << compressed_buffer_length <<
+      ")!");
+    // An error indicator is set by fread, so the following check will throw.
+  }
+
+  if (ferror(file_pointer)) {
+    fclose(file_pointer);
+
+    std::stringstream errmsg;
+    errmsg << "Unable to read binary data from file: \"" << uri << "\"!";
+
+    throw std::runtime_error(errmsg.str());
+  }
+
+  fclose(file_pointer);
+  return compressed_buffer;
+}
+
+void write_output_buffer(
+  const uint8_t * output_buffer,
+  const size_t output_buffer_length,
+  const std::string & uri)
+{
+  const auto file_pointer = open_file(uri.c_str(), "wb");
+  if (file_pointer == nullptr) {
+    std::stringstream errmsg;
+    errmsg << "Failed to open file: \"" << uri << "\" for binary writing!";
+
+    throw std::runtime_error(errmsg.str());
+  }
+
+  const auto nWrite = fwrite(
+    output_buffer, sizeof(uint8_t), output_buffer_length, file_pointer);
+
+  if (nWrite != output_buffer_length) {
+    ROSBAG2_COMPRESSION_LOG_ERROR_STREAM("Bytes written (" <<
+      nWrite << " != output_buffer_length (" << output_buffer_length <<
+      ")!");
+    // An error indicator is set by fwrite, so the following check will throw.
+  }
+
+  if (ferror(file_pointer)) {
+    fclose(file_pointer);
+
+    std::stringstream errmsg;
+    errmsg << "Unable to write decompressed data to file: \"" << uri << "\"!";
+
+    throw std::runtime_error(errmsg.str());
+  }
+
+  fclose(file_pointer);
+}
+
+void throw_on_zstd_error(const size_t compression_result)
+{
+  if (ZSTD_isError(compression_result)) {
+    std::stringstream error;
+    error << "ZSTD decompression error: " << ZSTD_getErrorName(compression_result);
+
+    throw std::runtime_error(error.str());
+  }
+}
+
+void check_frame_content(const size_t frame_content)
+{
+  if (frame_content == ZSTD_CONTENTSIZE_ERROR) {
+    throw std::runtime_error("File not compressed with Zstd.");
```

Comment:
Is not being compressed with `Zstd` the only possibility that returns `ZSTD_CONTENTSIZE_ERROR`? Perhaps include "file could be corrupt", etc.

Reply:
Changed error message to just state that an error occurred on `ZSTD_CONTENTSIZE_ERROR` since ZStandard doesn't elaborate further why it failed.